### PR TITLE
Adding the BFSI Consent Management Service Module Implementation

### PR DESCRIPTION
**Pull request by Ashi1993**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/org.wso2.bfsi.consent.management.dao/pom.xml [hunk 1]
```diff
@@ -204,11 +205,38 @@
...
+                            org.osgi.framework;version="${osgi.framework.imp.pkg.version.range}",
+                            org.osgi.service.component;version="${osgi.service.component.imp.pkg.version.range}",
+                            org.wso2.bfsi.consent.management.common.*;version="${project.version}",
+                            org.apache.commons.lang3;version="${commons-lang.version}"
+                        </Import-Package>
+                        <Export-Package>
+                            !org.wso2.bfsi.consent.management.dao.internal,
+                            org.wso2.bfsi.consent.management.dao.*;version="${project.version}",
+                        </Export-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (RivinduM) [inline components/org.wso2.bfsi.consent.management.dao/pom.xml]: Shall we remove the DynamicImport-Package and add any missing imports if there's any? Check other places also.

Comment (Ashi1993) [reply components/org.wso2.bfsi.consent.management.dao/pom.xml]: Fixed in https://github.com/wso2-extensions/bfsi-consent-management/pull/3/commits/c939420493a35062265681f22ab76c8f10ecc313

File: components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/internal/ConsentManagementServiceComponent.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            if (!isConnectionActive) {
+                log.error("The connection is not active");
+                throw new ConsentManagementRuntimeException("The connection is not active");
+            }
+        } catch (SQLException e) {
+            log.error("Database connection is not active, cannot proceed");
+            throw new ConsentManagementRuntimeException("Database connection is not active, cannot proceed");
+        }
+
+
```
Comment (RivinduM) [inline components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/internal/ConsentManagementServiceComponent.java]:
```suggestion

```

File: components/org.wso2.bfsi.consent.management.dao/src/main/java/org/wso2/bfsi/consent/management/dao/internal/ConsentManagementDaoServiceComponent.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Activate
+    protected void activate(ComponentContext context) {
+
+        log.debug("Consent Management DAO is registered successfully.");
+    }
+
+    @Deactivate
+    protected void deactivate(ComponentContext context) {
+        log.debug("Consent Management DAO is deactivated");
```
Comment (RivinduM) [inline components/org.wso2.bfsi.consent.management.dao/src/main/java/org/wso2/bfsi/consent/management/dao/internal/ConsentManagementDaoServiceComponent.java:46]:
```suggestion

        log.debug("Consent Management DAO is deactivated");
```

File: components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/util/ConsentCoreServiceUtil.java [hunk 4]
```diff
@@ -0,0 +1,738 @@
...
+    public static void postStateChange(Connection connection, ConsentCoreDAO consentCoreDAO, String consentID,
+                                       String userID, String newConsentStatus, String previousConsentStatus,
+                                       String reason, String clientId, Map<String, Object> consentDataMap)
+            throws ConsentDataInsertionException, ConsentManagementException {
+
+        createAuditRecord(connection, consentCoreDAO, consentID, userID, newConsentStatus, previousConsentStatus,
+                reason);
+        // TODO:
+//        ConsentStateChangeListenerImpl.getInstance().onStateChange(consentID, userID, newConsentStatus,
+//                previousConsentStatus, reason, clientId, consentDataMap);
```
Comment (RivinduM) [inline components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/util/ConsentCoreServiceUtil.java]: Let's remove the commented lines if not required. Else let's add a proper todo comment

**Review by RivinduM — DISMISSED** [review]

File: components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/impl/ConsentCoreServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,1905 @@
...
+                    String previousConsentStatus = resource.getCurrentStatus();
+
+                    // Update consent status
+                    if (log.isDebugEnabled()) {
+                        log.debug(String.format("Updating consent status for consent ID: %s",
+                                resource.getConsentID().replaceAll("[\r\n]", "")));
+                    }
+                    consentCoreDAO.updateConsentStatus(connection, resource.getConsentID(), revokedConsentStatus);
+
+                    //TODO
```
Comment (hasithakn) [inline components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/impl/ConsentCoreServiceImpl.java]: we can remove `todo` here right?

File: pom.xml [hunk 6]
```diff
@@ -272,20 +301,29 @@
...
-        <json.wso2.version>3.0.0.wso2v4</json.wso2.version>
-        <com.googlecode.json.version>1.1.1</com.googlecode.json.version>
+        <json-smart.version>2.4.10</json-smart.version>
         <commons-collections.version>4.4</commons-collections.version>
         <org.testng.version>7.7.0</org.testng.version>
         <jacoco.version>0.8.6</jacoco.version>
         <commons-dbcp.version>1.4</commons-dbcp.version>
         <orbit.version.h2.engine>1.2.140.wso2v3</orbit.version.h2.engine>
         <mockito.version>1.10.19</mockito.version>
         <powermock.version>1.7.1</powermock.version>
```
Comment (hasithakn) [inline pom.xml]: <powermock.version> is already defined, shall we remove this one?
```suggestion
```

Comment (hasithakn) [reply pom.xml]: Check mockito.version as well

Comment (Ashi1993) [reply pom.xml]: Fixed in https://github.com/wso2-extensions/bfsi-consent-management/pull/3/commits/08d281c66fc90f979dcae36a8c8baedf99ab9aa9

File: components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/impl/ConsentCoreServiceImpl.java (hunk 5 above)
Comment (Ashi1993) [reply components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/impl/ConsentCoreServiceImpl.java]: Fixed in https://github.com/wso2-extensions/bfsi-consent-management/pull/3/commits/40acdc39fe19861bf4e0a72e49bc8aa4f3e1e438

File: components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/util/ConsentCoreServiceUtil.java (hunk 4 above)
Comment (Ashi1993) [reply components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/util/ConsentCoreServiceUtil.java]: Fixed in https://github.com/wso2-extensions/bfsi-consent-management/pull/3/commits/08d281c66fc90f979dcae36a8c8baedf99ab9aa9

File: components/org.wso2.bfsi.consent.management.dao/src/main/java/org/wso2/bfsi/consent/management/dao/internal/ConsentManagementDaoServiceComponent.java (hunk 3 above)
Comment (Ashi1993) [reply components/org.wso2.bfsi.consent.management.dao/src/main/java/org/wso2/bfsi/consent/management/dao/internal/ConsentManagementDaoServiceComponent.java:46]: Fixed in https://github.com/wso2-extensions/bfsi-consent-management/pull/3/commits/40acdc39fe19861bf4e0a72e49bc8aa4f3e1e438

File: components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/internal/ConsentManagementServiceComponent.java (hunk 2 above)
Comment (Ashi1993) [reply components/org.wso2.bfsi.consent.management.service/src/main/java/org/wso2/bfsi/consent/management/service/internal/ConsentManagementServiceComponent.java]: Fixed in https://github.com/wso2-extensions/bfsi-consent-management/pull/3/commits/40acdc39fe19861bf4e0a72e49bc8aa4f3e1e438

**Review by hasithakn — APPROVED** [review]

**Review by RivinduM — APPROVED** [review]

**Pull request merged**